### PR TITLE
Codechange: Replace GroupStatistics' num_engines with map.

### PR DESCRIPTION
**Pull request by PeterN**
## Motivation / Problem

GroupStatistics has a manually memory managed array to count the number of engines of each type.

## Description

Replace with a map. This removes manual memory management with calloc/free calls, and allows only the needed entries to be populated, instead of an entry for every possible engine ID.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

**Review by TrueBrain — DISMISSED** [review]

File: src/group.h [hunk 1]
```diff
@@ -24,14 +24,13 @@ extern GroupPool _group_pool; ///< Pool of groups.
 struct GroupStatistics {
 	Money profit_last_year;                 ///< Sum of profits for all vehicles.
 	Money profit_last_year_min_age;         ///< Sum of profits for vehicles considered for profit statistics.
-	uint16_t *num_engines;                    ///< Caches the number of engines of each type the company owns.
+	std::vector<uint16_t> num_engines; ///< Caches the number of engines of each type the company owns.
```
Comment (glx22) [inline src/group.h]: All these comment not aligned :)

Comment (TrueBrain) [reply src/group.h]: Good! Let's stop trying to align them! :D

**Comment by LordAro** [comment]
A vector that is just immediately resized to the maximum doesn't feel like good usage to me

**Comment by PeterN** [comment]
> A vector that is just immediately resized to the maximum doesn't feel like good usage to me

It's possible a std::map would be a better fit here, presumably most groups have only a few types of engine. They'll never have all types...

**Comment by PeterN** [comment]
Now using a std::map instead, so nothing is allocated initially, and only engines used in the group are counted.

`GetNumEngines()` exists to prevent making a new entry when we only need to find a value.

File: src/group_cmd.cpp [hunk 2]
```diff
@@ -48,8 +38,19 @@ void GroupStatistics::Clear()
 	this->profit_last_year_min_age = 0;
 
 	/* This is also called when NewGRF change. So the number of engines might have changed. Reallocate. */
```
Comment (TrueBrain) [inline src/group_cmd.cpp]: Comment is now out-of-date.

**Review by TrueBrain — APPROVED** [review]

**Pull request merged**